### PR TITLE
Extend evaluated decision event with decisionVersion for evaluted decisions

### LINKED ISSUE #8867: Add version to the evaluated decisions referred from a evaluated decision record
**Is your feature request related to a problem? Please describe.**
Similar to #8643, we've discovered the need for the version of the evaluated decisions of an EvaluatedDecisionRecord.

EvaluatedDecisionRecord already contains a `decisionVersion`. But, decisions can be individually updated by referring to them in different DRGs. So, while the `decisionVersion` of the evaluated decision is useful, it does not inform log consumers of the version of the evaluated required decisions.

This information is useful to Operate for filtering and sorting of the evaluated decisions data.

For more details, please read the [slack discussion](https://camunda.slack.com/archives/C02JAMSFZS8/p1646141911782899).

**Describe the solution you'd like**
Add `decisionVersion` to `evaluatedDecisions` of `EvaluatedDecisionRecord`.

Example:
```
{
    "decisionKey":1,
    "decisionId":"decision-id",
    "decisionName":"decision-name",
    "decisionVersion":1,
    "decisionRequirementsKey":2,
    "decisionRequirementsId":"decision-requirements-id",
    "decisionOutput":'"decision-output"',
    "processDefinitionKey":3,
    "bpmnProcessId":"bpmn-process-id",
    "processInstanceKey":4,
    "elementInstanceKey":5,
    "elementId":"element-id",
    "evaluatedDecisions":[
        {
            "decisionId":"decision-id",
            "decisionName":"decision-name",
            "decisionKey":6,
            "decisionVersion": 1,
            "decisionOutput":'"decision-output"',
            "decisionType":"DECISION_TABLE",
            "evaluatedInputs":[
                {
                    "inputId":"input-id",
                    "inputName":"input-name",
                    "inputValue":'"input-value"'
                }
            ],
            "matchedRules":[
                {
                    "ruleId":"rule-id",
                    "ruleIndex":1,
                    "evaluatedOutputs":[
                        {
                            "outputId":"output-id",
                            "outputName":"output-name",
                            "outputValue":'"output-value"'
                        }
                    ]
                }
            ]
        }
    ],
    "evaluationFailureMessage":"evaluation-failure-message",
    "failedDecisionId":"failed-decision-id"
}
```

**Describe alternatives you've considered**
The alternative is that this would be solved on Operate's side by consuming the deployment records. This adds additional logic to the importer which slows it down. 

Of course, the same could be said for Zeebe, but Zeebe is much less slowed down by this than Operate. In addition, the data is already available for Zeebe at runtime. The only part that slows things down is the additional data that will be written in the records. Considering we also found this reasonable for `decisionKey` in #8643, I don't see why we can't also add `decisionVersion`. 

The same can be said for increasing the record size, which does grow a bit, but not by an unreasonable amount. 

**Additional context**
A similar change was done before in #8829, we can practically repeat these changes.

### PR DESCRIPTION
**Pull request by korthout**
## Description

This extends the EvaluatedDecision event by adding the `decisionVersion` to each of the `evaluatedDecisions`.

It also adds this to the Elasticsearch template so the ES exporter can export this property.

This PR is greatly inspired by https://github.com/camunda-cloud/zeebe/pull/8829, which did the same for the `decisionKey` of the `evaluatedDecisions`. Each commit describes which specific commit it was inspired by. This should help make the review easier.

## Related issues

closes #8867

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

### CODE REVIEW

**Comment by korthout** [comment]
I've manually tested it using the [Inspect data in Elasticsearch guide](https://github.com/camunda-cloud/zeebe/wiki/Inspect-data-in-Elasticsearch).

- first I deployed a drg that contains the `force-user` and the `jedi-or-sith` decisions.
- I also deployed a bpmn with a business rule task that refers to the `jedi-or-sith` decision
- then I created an instance of it and inspected the EvaluatedDecision event in Kibana
- we expect to see an evaluated decision with a single evaluatedDecisions entry
```
zbctl --insecure deploy ../examples/force-user.dmn
zbctl --insecure deploy ../examples/jedi-or-sith.bpmn
zbctl --insecure create instance jedi-or-sith
```

```
{
    "partitionId": 1,
    "value": {
      "decisionVersion": 1,
      "decisionOutput": "\"Jedi\"",
      "evaluatedDecisions": [
        {
          "decisionVersion": 1,
          "decisionOutput": "\"Jedi\"",
          "decisionType": "DECISION_TABLE",
          "evaluatedInputs": [
            {
              "inputId": "Input_1",
              "inputName": "Lightsaber color",
              "inputValue": "\"blue\""
            }
          ],
          "matchedRules": [
            {
              "ruleId": "DecisionRule_116a3qk",
              "ruleIndex": 4,
              "evaluatedOutputs": [
                {
                  "outputId": "Output_1",
                  "outputName": "jedi_or_sith",
                  "outputValue": "\"Jedi\""
                }
              ]
            }
          ],
          "decisionId": "jedi_or_sith",
          "decisionName": "Jedi or Sith",
          "decisionKey": 2251799813685250
        }
      ],
      "evaluationFailureMessage": "",
      "failedDecisionId": "",
      "decisionRequirementsId": "force_users",
      "decisionRequirementsKey": 2251799813685249,
      "decisionId": "jedi_or_sith",
      "decisionName": "Jedi or Sith",
      "decisionKey": 2251799813685250,
      "processDefinitionKey": 2251799813685253,
      "bpmnProcessId": "jedi-or-sith",
      "processInstanceKey": 2251799813685255,
      "elementInstanceKey": 2251799813685259,
      "elementId": "Activity_0bsn2q0"
    },
    "key": 2251799813685261,
    "timestamp": 1646151501694,
    "sourceRecordPosition": 23,
    "intent": "EVALUATED",
    "rejectionType": "NULL_VAL",
    "rejectionReason": "",
    "brokerVersion": "1.4.0",
    "valueType": "DECISION_EVALUATION",
    "recordType": "EVENT",
    "position": 26
  }
```

- next I deployed another bpmn that contains a business rule task that calls the other decision
- and I created an instance of it, and looked at the data in Kibana
- we now expect an evaluated decision record with 2 evaluatedDecisions entries.

```
zbctl --insecure deploy ../examples/force-user.bpmn
zbctl --insecure create instance jedi-or-sith.bpmn
```

```
{
    "partitionId": 1,
    "value": {
      "decisionVersion": 1,
      "decisionOutput": "\"Obi-Wan Kenobi\"",
      "evaluatedDecisions": [
        {
          "decisionVersion": 1,
          "decisionOutput": "\"Jedi\"",
          "decisionType": "DECISION_TABLE",
          "evaluatedInputs": [
            {
              "inputId": "Input_1",
              "inputName": "Lightsaber color",
              "inputValue": "\"blue\""
            }
          ],
          "matchedRules": [
            {
              "ruleId": "DecisionRule_0zumznl",
              "ruleIndex": 1,
              "evaluatedOutputs": [
                {
                  "outputId": "Output_1",
                  "outputName": "jedi_or_sith",
                  "outputValue": "\"Jedi\""
                }
              ]
            }
          ],
          "decisionId": "jedi_or_sith",
          "decisionName": "Jedi or Sith",
          "decisionKey": 2251799813685250
        },
        {
          "decisionVersion": 1,
          "decisionOutput": "\"Obi-Wan Kenobi\"",
          "decisionType": "DECISION_TABLE",
          "evaluatedInputs": [
            {
              "inputId": "InputClause_0qnqj25",
              "inputName": "Jedi or Sith",
              "inputValue": "\"Jedi\""
            },
            {
              "inputId": "InputClause_0k64hys",
              "inputName": "Body height",
              "inputValue": "185"
            }
          ],
          "matchedRules": [
            {
              "ruleId": "DecisionRule_0uin2hk",
              "ruleIndex": 2,
              "evaluatedOutputs": [
                {
                  "outputId": "OutputClause_0hhe1yo",
                  "outputName": "force_user",
                  "outputValue": "\"Obi-Wan Kenobi\""
                }
              ]
            }
          ],
          "decisionId": "force_user",
          "decisionName": "Which force user?",
          "decisionKey": 2251799813685251
        }
      ],
      "evaluationFailureMessage": "",
      "failedDecisionId": "",
      "decisionRequirementsId": "force_users",
      "decisionRequirementsKey": 2251799813685249,
      "decisionId": "force_user",
      "decisionName": "Which force user?",
      "decisionKey": 2251799813685251,
      "processDefinitionKey": 2251799813685310,
      "bpmnProcessId": "force-user",
      "processInstanceKey": 2251799813685312,
      "elementInstanceKey": 2251799813685316,
      "elementId": "Activity_1sl5qf7"
    },
    "key": 2251799813685319,
    "timestamp": 1646151999424,
    "sourceRecordPosition": 159,
    "intent": "EVALUATED",
    "rejectionType": "NULL_VAL",
    "rejectionReason": "",
    "brokerVersion": "1.4.0",
    "valueType": "DECISION_EVALUATION",
    "recordType": "EVENT",
    "position": 163
  }
```

- next I deployed another drg that only contains the `jedi-or-sith` decision 
  this should create a decision, with a new key, and new version but with the same decisionId
- I then created an instance of the jedi-and-sith process and checked the event in Kibana
- we expect that it shows `decisionVersion: 2` in both the record as well as the `evaluatedDecisions` property

```
zbctl --insecure deploy ../examples/jedi-or-sith.dmn
zbctl --insecure create instance jedi-or-sith
```

```
{
    "partitionId": 1,
    "value": {
      "decisionVersion": 2,
      "decisionOutput": "\"Jedi\"",
      "evaluatedDecisions": [
        {
          "decisionVersion": 2,
          "decisionOutput": "\"Jedi\"",
          "decisionType": "DECISION_TABLE",
          "evaluatedInputs": [
            {
              "inputId": "Input_1",
              "inputName": "Lightsaber color",
              "inputValue": "\"blue\""
            }
          ],
          "matchedRules": [
            {
              "ruleId": "DecisionRule_116a3qk",
              "ruleIndex": 4,
              "evaluatedOutputs": [
                {
                  "outputId": "Output_1",
                  "outputName": "jedi_or_sith",
                  "outputValue": "\"Jedi\""
                }
              ]
            }
          ],
          "decisionId": "jedi_or_sith",
          "decisionName": "Jedi or Sith",
          "decisionKey": 2251799813685327
        }
      ],
      "evaluationFailureMessage": "",
      "failedDecisionId": "",
      "decisionRequirementsId": "jedi_or_siths",
      "decisionRequirementsKey": 2251799813685326,
      "decisionId": "jedi_or_sith",
      "decisionName": "Jedi or Sith",
      "decisionKey": 2251799813685327,
      "processDefinitionKey": 2251799813685253,
      "bpmnProcessId": "jedi-or-sith",
      "processInstanceKey": 2251799813685329,
      "elementInstanceKey": 2251799813685333,
      "elementId": "Activity_0bsn2q0"
    },
    "key": 2251799813685335,
    "timestamp": 1646152072399,
    "sourceRecordPosition": 202,
    "intent": "EVALUATED",
    "rejectionType": "NULL_VAL",
    "rejectionReason": "",
    "brokerVersion": "1.4.0",
    "valueType": "DECISION_EVALUATION",
    "recordType": "EVENT",
    "position": 205
  }
```

- finally I also created an instance of the other process
- and checked the event in kibana
- we expect to see an event referring to version 1 of the decisions

```
{
    "partitionId": 1,
    "value": {
      "decisionVersion": 1,
      "decisionOutput": "\"Obi-Wan Kenobi\"",
      "evaluatedDecisions": [
        {
          "decisionVersion": 1,
          "decisionOutput": "\"Jedi\"",
          "decisionType": "DECISION_TABLE",
          "evaluatedInputs": [
            {
              "inputId": "Input_1",
              "inputName": "Lightsaber color",
              "inputValue": "\"blue\""
            }
          ],
          "matchedRules": [
            {
              "ruleId": "DecisionRule_0zumznl",
              "ruleIndex": 1,
              "evaluatedOutputs": [
                {
                  "outputId": "Output_1",
                  "outputName": "jedi_or_sith",
                  "outputValue": "\"Jedi\""
                }
              ]
            }
          ],
          "decisionId": "jedi_or_sith",
          "decisionName": "Jedi or Sith",
          "decisionKey": 2251799813685250
        },
        {
          "decisionVersion": 1,
          "decisionOutput": "\"Obi-Wan Kenobi\"",
          "decisionType": "DECISION_TABLE",
          "evaluatedInputs": [
            {
              "inputId": "InputClause_0qnqj25",
              "inputName": "Jedi or Sith",
              "inputValue": "\"Jedi\""
            },
            {
              "inputId": "InputClause_0k64hys",
              "inputName": "Body height",
              "inputValue": "185"
            }
          ],
          "matchedRules": [
            {
              "ruleId": "DecisionRule_0uin2hk",
              "ruleIndex": 2,
              "evaluatedOutputs": [
                {
                  "outputId": "OutputClause_0hhe1yo",
                  "outputName": "force_user",
                  "outputValue": "\"Obi-Wan Kenobi\""
                }
              ]
            }
          ],
          "decisionId": "force_user",
          "decisionName": "Which force user?",
          "decisionKey": 2251799813685251
        }
      ],
      "evaluationFailureMessage": "",
      "failedDecisionId": "",
      "decisionRequirementsId": "force_users",
      "decisionRequirementsKey": 2251799813685249,
      "decisionId": "force_user",
      "decisionName": "Which force user?",
      "decisionKey": 2251799813685251,
      "processDefinitionKey": 2251799813685310,
      "bpmnProcessId": "force-user",
      "processInstanceKey": 2251799813685343,
      "elementInstanceKey": 2251799813685347,
      "elementId": "Activity_1sl5qf7"
    },
    "key": 2251799813685350,
    "timestamp": 1646152101455,
    "sourceRecordPosition": 240,
    "intent": "EVALUATED",
    "rejectionType": "NULL_VAL",
    "rejectionReason": "",
    "brokerVersion": "1.4.0",
    "valueType": "DECISION_EVALUATION",
    "recordType": "EVENT",
    "position": 244
  }
```

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnDecisionBehavior.java [hunk 1]
```diff
@@ -317,4 +318,10 @@ private void addOutputToEvaluationEvent(
         .setOutputName(evaluatedOutput.outputName())
         .setOutputValue(evaluatedOutput.outputValue());
   }
+
+  private record DecisionInfo(long key, int version) {
+    DecisionInfo(final PersistedDecision persistedDecision) {
+      this(persistedDecision.getDecisionKey(), persistedDecision.getVersion());
+    }
+  }
```
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnDecisionBehavior.java:326]: Oehhh fancy 👍 
Github syntax highlighting doesn't know `record` yet 😄

**Review by remcowesterhoud — APPROVED** [review]
Thanks @korthout, the code seems fine 🚀 

❓ One thing I don't fully understand is how the `descisionVersion` of the EvaluatedDecisionRecord could differ from the `decisionVersion` of the `evaluatedDecision` in this record. I know you've said:

> EvaluatedDecisionRecord already contains a decisionVersion. But, decisions can be individually updated by referring to them in different DRGs. So, while the decisionVersion of the evaluated decision is useful, it does not inform log consumers of the version of the evaluated required decisions.

So I figure you know it can differ. Care to explain this to me sometime?

**Comment by korthout** [comment]
> One thing I don't fully understand is how the `descisionVersion` of the EvaluatedDecisionRecord could differ from the `decisionVersion` of the `evaluatedDecision` in this record

@remcowesterhoud Sure! Let me give it a try:

TBH I was not sure about it either, but I asked @sdorokhova and she confirmed that the decision version of a required decision could differ from the decision version of the evaluated decision. After my manual test did not show the versions diverging, it was on my mind for some time longer and I ended up running an additional manual test that confirmed that they can differ.

The simplest case where they differ is by first deploying a DMN with a single decision. Next, you add a required decision to the decision in the DRG and you deploy this new version of the DMN. Doing so results in a new version of the DRG (v2), a new version of the root decision (v2) and a new decision (v1) which is the required decision of the root decision. 

Does that help?

**Comment by korthout** [comment]
Also, let's merge this

bors merge

**Comment by remcowesterhoud** [comment]
@korthout That does help. Thanks! 🙇

**Pull request merged**